src/docfx/build/dependency/DependencyMapBuilder.cs

```diff
@@ -34,13 +35,52 @@ public void AddDependencyItem(Document from, Document to, DependencyType type)
 
         public DependencyMap Build()
         {
-            return new DependencyMap(_dependencyItems
-                .GroupBy(k => k.Source)
+            return new DependencyMap(Flatten());
+        }
+
+        private Dictionary<Document, HashSet<DependencyItem>> Flatten()
+        {
+            var result = new Dictionary<Document, HashSet<DependencyItem>>();
+            var graph = _dependencyItems
+                .GroupBy(k => k.From)
                 .ToDictionary(
                     k => k.Key,
-                    v => (from r in v.Distinct()
-                         orderby r.Dest.FilePath, r.Type
-                         select r).ToList()));
+                    v => (from r in v
+                          orderby r.To.FilePath, r.Type
+                          select r).ToHashSet());
```

Comment:
you need cache the calculated nodes, otherwise a lot of nodes will be re-calculated multiple times.

like that ?
```C# 
flattenCache

GetFlattenItems(key, graph, items)
{
    return flattenCache.GetOrAdd(key, => 
    {
        var flattenItems = items.ToList();
        foreach (var item in items)
        {
            if (graph.ContainsKey(item) && item != key && CanTransit)
            {
                flattenItems.AddRange(GetFlattenItems(item, graph, graph[item]))
            }
        }
        return flattenItems;
    });
}

var flattenGrap = graph.ToDictionary();
foreach(var (from, items) in graph)
{
    flattenGrap[from] = GetFlattenItems(graph, items);
}
```

Reply:
good point, we can reuse the calculated result